### DISCUSSION TITLE
Can someone please explain saved_hybrid?

### DISCUSSION THREAD

**Discussion by halo34** [ultralytics/ultralytics#8761] [Q&A] [answered]
Hi,
Val has the option saved_hybrid. 
But I don't understand what saved_hybrid should do.
If I activate the option, the validation images show me only the ground truth with 1.0, not the predicted label. 
The ground trues are also used for the statistics, but as far as I understand, they should only be based on the model's predictions.
Therefore, what should the flag do?

**Comment by glenn-jocher** [comment]
@halo34 hi there! 👋

The `saved_hybrid` option during validation creates a hybrid version of your labels that combines both ground truth and model predictions. When activated, it's expected to see ground truth labels with a confidence of 1.0, as this feature is designed to help visualize how well the model's predictions align with the actual ground truth. The primary purpose is for easier inspection and debugging of model performance visually.

However, the statistics should indeed be based on the model's predictions compared to the ground truth, not the ground truth itself. If you're seeing only ground truth used for statistics, that might not be the intended behavior. The flag is meant to enhance understanding of model performance through visual inspection alongside traditional metrics.

Here's a quick example of how you might use it:
```bash
yolo val model=yolov8n.pt data=coco128.yaml save_hybrid=True
```

This will save images annotated with both predictions (if any) and ground truth labels for comparison. Hope this clarifies things! 😊

**Reply by halo34** [reply]
Hi, 
Thank you very much for the answer.
However, the graphics are still different. For example, the f1 score of Coco8 validation. So should you just use save_hybrid=false for the metrics?
save_hybrid=False:
<img width="631" alt="Bildschirmfoto 2024-03-08 um 10 33 30" src="https://github.com/ultralytics/ultralytics/assets/49497711/6f4753a5-4e4a-417f-84ff-353d9cb29402">

and with save_hybrid=True:
<img width="637" alt="Bildschirmfoto 2024-03-08 um 10 33 19" src="https://github.com/ultralytics/ultralytics/assets/49497711/cff27447-8ef6-4561-ada9-f18454588808">

**Reply by glenn-jocher** [reply, marked as answer]
@halo34 hey there! 🌟

Thanks for bringing this to our attention. The `save_hybrid` option is primarily for visualization purposes, allowing you to see both ground truth and predictions overlaid on your validation images. It's a great tool for a quick visual check on how your model is performing.

For accurate metrics like the F1 score, you're right; it's best to use `save_hybrid=False`. This ensures that the metrics are calculated purely based on the model's predictions against the ground truth, without any overlay confusion.

Here's a quick snippet for clarity:
```bash
yolo val model=yolov8n.pt data=coco128.yaml save_hybrid=False
```

This command will give you the clean metrics for model evaluation. Hope this clears things up! Happy modeling! 😊

**Reply by akriegler** [reply]
So what is the point of it? You can either get visualizations, but wrong metrics, or no visualizations and correct metrics?

**Reply by glenn-jocher** [reply]
The `save_hybrid=True` option is specifically designed for **visual debugging** by overlaying ground truth labels onto prediction visuals, but as noted in the [warning in our Val documentation](https://docs.ultralytics.com/modes/val/), it intentionally modifies label comparisons and should not be used for metric evaluation. For accurate performance metrics, always use `save_hybrid=False` (default). These are separate workflows: use hybrid mode for visual analysis, and standard validation for metric calculations.
